// @milkdown/design-system/lib/index.es.js
var R = (r, e, t) => {
  if (!e.has(r))
    throw TypeError("Cannot " + t);
};
var u = (r, e, t) => (R(r, e, "read from private field"), t ? t.call(r) : e.get(r)), b = (r, e, t) => {
  if (e.has(r))
    throw TypeError("Cannot add the same private member more than once");
  e instanceof WeakSet ? e.add(r) : e.set(r, t);
}, A = (r, e, t, n) => (R(r, e, "write to private field"), n ? n.call(r, t) : e.set(r, t), t);
import U from "@emotion/cache";
import { createSlice as S, createContainer as W } from "@milkdown/ctx";
function q(r) {
  for (var e = 0, t, n = 0, s = r.length; s >= 4; ++n, s -= 4)
    t = r.charCodeAt(n) & 255 | (r.charCodeAt(++n) & 255) << 8 | (r.charCodeAt(++n) & 255) << 16 | (r.charCodeAt(++n) & 255) << 24, t = (t & 65535) * 1540483477 + ((t >>> 16) * 59797 << 16), t ^= t >>> 24, e = (t & 65535) * 1540483477 + ((t >>> 16) * 59797 << 16) ^ (e & 65535) * 1540483477 + ((e >>> 16) * 59797 << 16);
  switch (s) {
    case 3:
      e ^= (r.charCodeAt(n + 2) & 255) << 16;
    case 2:
      e ^= (r.charCodeAt(n + 1) & 255) << 8;
    case 1:
      e ^= r.charCodeAt(n) & 255, e = (e & 65535) * 1540483477 + ((e >>> 16) * 59797 << 16);
  }
  return e ^= e >>> 13, e = (e & 65535) * 1540483477 + ((e >>> 16) * 59797 << 16), ((e ^ e >>> 15) >>> 0).toString(36);
}
var B = {
  animationIterationCount: 1,
  borderImageOutset: 1,
  borderImageSlice: 1,
  borderImageWidth: 1,
  boxFlex: 1,
  boxFlexGroup: 1,
  boxOrdinalGroup: 1,
  columnCount: 1,
  columns: 1,
  flex: 1,
  flexGrow: 1,
  flexPositive: 1,
  flexShrink: 1,
  flexNegative: 1,
  flexOrder: 1,
  gridRow: 1,
  gridRowEnd: 1,
  gridRowSpan: 1,
  gridRowStart: 1,
  gridColumn: 1,
  gridColumnEnd: 1,
  gridColumnSpan: 1,
  gridColumnStart: 1,
  msGridRow: 1,
  msGridRowSpan: 1,
  msGridColumn: 1,
  msGridColumnSpan: 1,
  fontWeight: 1,
  lineHeight: 1,
  opacity: 1,
  order: 1,
  orphans: 1,
  tabSize: 1,
  widows: 1,
  zIndex: 1,
  zoom: 1,
  WebkitLineClamp: 1,
  fillOpacity: 1,
  floodOpacity: 1,
  stopOpacity: 1,
  strokeDasharray: 1,
  strokeDashoffset: 1,
  strokeMiterlimit: 1,
  strokeOpacity: 1,
  strokeWidth: 1
};
function K(r) {
  var e = /* @__PURE__ */ Object.create(null);
  return function(t) {
    return e[t] === void 0 && (e[t] = r(t)), e[t];
  };
}
var I = `You have illegal escape sequence in your template literal, most likely inside content's property value.
Because you write your CSS inside a JavaScript string you actually have to do double escaping, so for example "content: '\\00d7';" should become "content: '\\\\00d7';".
You can read more about this here:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#ES2018_revision_of_illegal_escape_sequences`, J = "You have passed in falsy value as style object's key (can happen when in example you pass unexported component as computed key).", H = /[A-Z]|^ms/g, L = /_EMO_([^_]+?)_([^]*?)_EMO_/g, O = function(e) {
  return e.charCodeAt(1) === 45;
}, V = function(e) {
  return e != null && typeof e != "boolean";
}, N = /* @__PURE__ */ K(function(r) {
  return O(r) ? r : r.replace(H, "-$&").toLowerCase();
}), w = function(e, t) {
  switch (e) {
    case "animation":
    case "animationName":
      if (typeof t == "string")
        return t.replace(L, function(n, s, o) {
          return h = {
            name: s,
            styles: o,
            next: h
          }, s;
        });
  }
  return B[e] !== 1 && !O(e) && typeof t == "number" && t !== 0 ? t + "px" : t;
};
if (process.env.NODE_ENV !== "production") {
  var Q = /(var|attr|counters?|url|(((repeating-)?(linear|radial))|conic)-gradient)\(|(no-)?(open|close)-quote/, Z = ["normal", "none", "initial", "inherit", "unset"], X = w, k = /^-ms-/, ee = /-(.)/g, D = {};
  w = function(e, t) {
    if (e === "content" && (typeof t != "string" || Z.indexOf(t) === -1 && !Q.test(t) && (t.charAt(0) !== t.charAt(t.length - 1) || t.charAt(0) !== '"' && t.charAt(0) !== "'")))
      throw new Error("You seem to be using a value for 'content' without quotes, try replacing it with `content: '\"" + t + "\"'`");
    var n = X(e, t);
    return n !== "" && !O(e) && e.indexOf("-") !== -1 && D[e] === void 0 && (D[e] = !0, console.error("Using kebab-case for css properties in objects is not supported. Did you mean " + e.replace(k, "ms-").replace(ee, function(s, o) {
      return o.toUpperCase();
    }) + "?")), n;
  };
}
var z = "Component selectors can only be used in conjunction with @emotion/babel-plugin, the swc Emotion plugin, or another Emotion-aware compiler transform.";
function x(r, e, t) {
  if (t == null)
    return "";
  if (t.__emotion_styles !== void 0) {
    if (process.env.NODE_ENV !== "production" && t.toString() === "NO_COMPONENT_SELECTOR")
      throw new Error(z);
    return t;
  }
  switch (typeof t) {
    case "boolean":
      return "";
    case "object": {
      if (t.anim === 1)
        return h = {
          name: t.name,
          styles: t.styles,
          next: h
        }, t.name;
      if (t.styles !== void 0) {
        var n = t.next;
        if (n !== void 0)
          for (; n !== void 0; )
            h = {
              name: n.name,
              styles: n.styles,
              next: h
            }, n = n.next;
        var s = t.styles + ";";
        return process.env.NODE_ENV !== "production" && t.map !== void 0 && (s += t.map), s;
      }
      return te(r, e, t);
    }
    case "function": {
      if (r !== void 0) {
        var o = h, i = t(r);
        return h = o, x(r, e, i);
      } else
        process.env.NODE_ENV !== "production" && console.error("Functions that are interpolated in css calls will be stringified.\nIf you want to have a css call based on props, create a function that returns a css call like this\nlet dynamicStyle = (props) => css`color: ${props.color}`\nIt can be called directly with props or interpolated in a styled call like this\nlet SomeComponent = styled('div')`${dynamicStyle}`");
      break;
    }
    case "string":
      if (process.env.NODE_ENV !== "production") {
        var c = [], f = t.replace(L, function(a, m, p) {
          var T = "animation" + c.length;
          return c.push("const " + T + " = keyframes`" + p.replace(/^@keyframes animation-\w+/, "") + "`"), "${" + T + "}";
        });
        c.length && console.error("`keyframes` output got interpolated into plain string, please wrap it with `css`.\n\nInstead of doing this:\n\n" + [].concat(c, ["`" + f + "`"]).join(`
`) + `

You should wrap it with \`css\` like this:

` + ("css`" + f + "`"));
      }
      break;
  }
  if (e == null)
    return t;
  var l = e[t];
  return l !== void 0 ? l : t;
}
function te(r, e, t) {
  var n = "";
  if (Array.isArray(t))
    for (var s = 0; s < t.length; s++)
      n += x(r, e, t[s]) + ";";
  else
    for (var o in t) {
      var i = t[o];
      if (typeof i != "object")
        e != null && e[i] !== void 0 ? n += o + "{" + e[i] + "}" : V(i) && (n += N(o) + ":" + w(o, i) + ";");
      else {
        if (o === "NO_COMPONENT_SELECTOR" && process.env.NODE_ENV !== "production")
          throw new Error(z);
        if (Array.isArray(i) && typeof i[0] == "string" && (e == null || e[i[0]] === void 0))
          for (var c = 0; c < i.length; c++)
            V(i[c]) && (n += N(o) + ":" + w(o, i[c]) + ";");
        else {
          var f = x(r, e, i);
          switch (o) {
            case "animation":
            case "animationName": {
              n += N(o) + ":" + f + ";";
              break;
            }
            default:
              process.env.NODE_ENV !== "production" && o === "undefined" && console.error(J), n += o + "{" + f + "}";
          }
        }
      }
    }
  return n;
}
var M = /label:\s*([^\s;\n{]+)\s*(;|$)/g, F;
process.env.NODE_ENV !== "production" && (F = /\/\*#\ssourceMappingURL=data:application\/json;\S+\s+\*\//g);
var h, C = function(e, t, n) {
  if (e.length === 1 && typeof e[0] == "object" && e[0] !== null && e[0].styles !== void 0)
    return e[0];
  var s = !0, o = "";
  h = void 0;
  var i = e[0];
  i == null || i.raw === void 0 ? (s = !1, o += x(n, t, i)) : (process.env.NODE_ENV !== "production" && i[0] === void 0 && console.error(I), o += i[0]);
  for (var c = 1; c < e.length; c++)
    o += x(n, t, e[c]), s && (process.env.NODE_ENV !== "production" && i[c] === void 0 && console.error(I), o += i[c]);
  var f;
  process.env.NODE_ENV !== "production" && (o = o.replace(F, function(p) {
    return f = p, "";
  })), M.lastIndex = 0;
  for (var l = "", a; (a = M.exec(o)) !== null; )
    l += "-" + a[1];
  var m = q(o) + l;
  return process.env.NODE_ENV !== "production" ? {
    name: m,
    styles: o,
    map: f,
    next: h,
    toString: function() {
      return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop).";
    }
  } : {
    name: m,
    styles: o,
    next: h
  };
}, ne = !0;
function $(r, e, t) {
  var n = "";
  return t.split(" ").forEach(function(s) {
    r[s] !== void 0 ? e.push(r[s] + ";") : n += s + " ";
  }), n;
}
var re = function(e, t, n) {
  var s = e.key + "-" + t.name;
  (n === !1 || ne === !1) && e.registered[s] === void 0 && (e.registered[s] = t.styles);
}, oe = function(e, t, n) {
  re(e, t, n);
  var s = e.key + "-" + t.name;
  if (e.inserted[t.name] === void 0) {
    var o = t;
    do
      e.insert(t === o ? "." + s : "", o, e.sheet, !0), o = o.next;
    while (o !== void 0);
  }
};
function G(r, e) {
  if (r.inserted[e.name] === void 0)
    return r.insert("", e, r.sheet, !0);
}
function j(r, e, t) {
  var n = [], s = $(r, n, t);
  return n.length < 2 ? t : s + e(n);
}
var se = function(e) {
  var t = U(e);
  t.sheet.speedy = function(c) {
    if (process.env.NODE_ENV !== "production" && this.ctr !== 0)
      throw new Error("speedy must be changed before any rules are inserted");
    this.isSpeedy = c;
  }, t.compat = !0;
  var n = function() {
    for (var f = arguments.length, l = new Array(f), a = 0; a < f; a++)
      l[a] = arguments[a];
    var m = C(l, t.registered, void 0);
    return oe(t, m, !1), t.key + "-" + m.name;
  }, s = function() {
    for (var f = arguments.length, l = new Array(f), a = 0; a < f; a++)
      l[a] = arguments[a];
    var m = C(l, t.registered), p = "animation-" + m.name;
    return G(t, {
      name: m.name,
      styles: "@keyframes " + p + "{" + m.styles + "}"
    }), p;
  }, o = function() {
    for (var f = arguments.length, l = new Array(f), a = 0; a < f; a++)
      l[a] = arguments[a];
    var m = C(l, t.registered);
    G(t, m);
  }, i = function() {
    for (var f = arguments.length, l = new Array(f), a = 0; a < f; a++)
      l[a] = arguments[a];
    return j(t.registered, n, ie(l));
  };
  return {
    css: n,
    cx: i,
    injectGlobal: o,
    keyframes: s,
    hydrate: function(f) {
      f.forEach(function(l) {
        t.inserted[l] = !0;
      });
    },
    flush: function() {
      t.registered = {}, t.inserted = {}, t.sheet.flush();
    },
    sheet: t.sheet,
    cache: t,
    getRegisteredStyles: $.bind(null, t.registered),
    merge: j.bind(null, t.registered, n)
  };
}, ie = function r(e) {
  for (var t = "", n = 0; n < e.length; n++) {
    var s = e[n];
    if (s != null) {
      var o = void 0;
      switch (typeof s) {
        case "boolean":
          break;
        case "object": {
          if (Array.isArray(s))
            o = r(s);
          else {
            o = "";
            for (var i in s)
              s[i] && i && (o && (o += " "), o += i);
          }
          break;
        }
        default:
          o = s;
      }
      o && (t && (t += " "), t += o);
    }
  }
  return t;
};
const ae = se, Se = (r) => ae(r), Ne = S({ key: "milkdown" }, "EmotionConfig"), _ = S({}, "Emotion"), d = (r) => S(() => null, r);
var g, v, E, y;
class Y {
  constructor() {
    b(this, g, W());
    b(this, v, /* @__PURE__ */ new Map());
    b(this, E, /* @__PURE__ */ new Set());
    b(this, y, () => {
    });
  }
  inject(e) {
    e(u(this, g).sliceMap);
  }
  has(e) {
    const t = typeof e == "string" ? e : e.sliceName;
    return u(this, v).has(t) ? !0 : u(this, g).getSlice(e).get()(null) != null;
  }
  set(e, t) {
    const n = typeof e == "string" ? e : e.sliceName;
    u(this, v).set(n, t);
  }
  get(e, t) {
    const n = typeof e == "string" ? e : e.sliceName, s = u(this, v).get(n);
    return s && (u(this, g).getSlice(e).set(s), u(this, v).delete(e)), u(this, g).getSlice(e).get()(t);
  }
  getSlice(e) {
    const t = typeof e == "string" ? e : e.sliceName, n = u(this, v).get(t);
    return n && (u(this, g).getSlice(e).set(n), u(this, v).delete(e)), u(this, g).getSlice(e).get();
  }
  onFlush(e, t = !0) {
    u(this, E).has(e) || u(this, E).add(e), t && e();
  }
  async switch(e, t) {
    e.get(_).flush(), await t(e)(e), this.runExecutor(), u(this, E).forEach((s) => s());
  }
  flush(e) {
    e.get(_).flush(), this.runExecutor(), u(this, E).forEach((n) => n());
  }
  setExecutor(e) {
    e(), A(this, y, e);
  }
  runExecutor() {
    u(this, y).call(this);
  }
}
g = new WeakMap(), v = new WeakMap(), E = new WeakMap(), y = new WeakMap();
const Ce = S(new Y(), "themeManager"), Oe = () => new Y(), ce = d(
  "border"
), P = d("color"), fe = d("font"), le = d("global"), ue = d("icon"), me = d("code-fence"), de = d("image"), he = d(
  "inner-editor"
), pe = d("input-chip"), ge = d("task-list-item"), ve = d("scrollbar"), Ee = d("shadow"), be = d("size"), Te = [
  P,
  be,
  fe,
  ve,
  Ee,
  ce,
  ue,
  le,
  me,
  de,
  he,
  ge,
  pe
], Re = (r) => {
  const e = /^([a-f\d])([a-f\d])([a-f\d])$/i, t = /^([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i, n = (l) => parseInt(l, 16), s = r.slice(1).replace(e, (l, a, m, p) => a + a + m + m + p + p), [o, i = "0", c = "0", f = "0"] = s.match(t) || [];
  return o ? [i, c, f].map(n) : null;
}, Ae = (r) => (e, t = 1) => r.get(P, [e, t]);
export {
  ce as ThemeBorder,
  me as ThemeCodeFence,
  P as ThemeColor,
  fe as ThemeFont,
  le as ThemeGlobal,
  ue as ThemeIcon,
  de as ThemeImage,
  he as ThemeInnerEditor,
  pe as ThemeInputChip,
  Y as ThemeManager,
  ve as ThemeScrollbar,
  Ee as ThemeShadow,
  be as ThemeSize,
  ge as ThemeTaskListItem,
  Oe as createThemeManager,
  d as createThemeSliceKey,
  Ne as emotionConfigCtx,
  _ as emotionCtx,
  Ae as getPalette,
  Re as hex2rgb,
  Se as initEmotion,
  Te as internalThemeKeys,
  Ce as themeManagerCtx
};
